### PR TITLE
add end to end diagnostic support for MQTT protocol

### PR DESCRIPTION
**Pull request by erich-wang**
1. Add option "diag_sampling_percentage" to support diagnostic if the value > 0
2. Add two system properties to support diagnostic info: diag-id & creation-time-utc

### CODE REVIEW

File: iothub_client/inc/iothub_client_diagnostic.h [hunk 1]
```diff
@@ -0,0 +1,47 @@
+// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+/** @file   iothub_client_diagnostic.h
+*	@brief  The @c diagnostic component.
```
Comment (ewertons) [inline iothub_client/inc/iothub_client_diagnostic.h]: That's good that you added this documentation, thanks.
Suggestion: could you add more details to give a better idea of what the module does? 'The diagnostic component' doesn't explain much.

File: iothub_client/inc/iothub_client_options.h [hunk 2]
```diff
@@ -43,6 +43,9 @@ extern "C"
     */
     static const char* OPTION_C2D_KEEP_ALIVE_FREQ_SECS = "c2d_keep_alive_freq_secs";
 
+    //diagnostic sampling percentage value, [0-100]
+    static const char* OPTION_DIAGNOSTIC_SAMPLING_PERCENTAGE = "diag_sampling_percentage";
```
Comment (ewertons) [inline iothub_client/inc/iothub_client_options.h]: What is the plan for the diagnostics? Will this eventually be applicable to all kinds of messages (Telemetry, Twin, ...)?

If this is restricted to Telemetry messages, this option should be more specific.

Comment (ewertons) [reply iothub_client/inc/iothub_client_options.h]: e.g., OPTION_TELEMETRY_DIAGNOSTIC_SAMPLING_PERCENTAGE = "telemetry_...

File: iothub_client/src/iothub_client_diagnostic.c [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+    }
+    return false;
+}
+
+bool IoTHubClient_Diagnostic_AddIfNecessary(IOTHUB_DIAGNOSTIC_SETTING_DATA* diagSetting, IOTHUB_MESSAGE_HANDLE messageHandle)
+{
+    /* Codes_SRS_IOTHUB_DIAGNOSTIC_13_001: [ IoTHubClient_Diagnostic_AddIfNecessary should return false if diagSetting or messageHandle is NULL. ]*/
+    if (diagSetting == NULL || messageHandle == NULL)
+    {
+        return false;
```
Comment (ewertons) [inline iothub_client/src/iothub_client_diagnostic.c]: the APIs in the C SDK follow a few guidelines that need to be implemented here:
1. Each function shall have only one point of return (create a variable on the top of the function, set it along the function and return its value at the end - see other functions in the C SDK to use as reference);
2. Functions in the C SDK usually return an int (see "__FAILURE__" in other functions in the C SDK)

File: iothub_client/src/iothub_client_diagnostic.c [hunk 4]
```diff
@@ -0,0 +1,103 @@
...
+    {
+        if (diagSetting->currentMessageNumber == UINT32_MAX)
+        {
+            diagSetting->currentMessageNumber %= diagSetting->diagSamplingPercentage;
+        }
+        ++diagSetting->currentMessageNumber;
+
+        double number = diagSetting->currentMessageNumber;
+        double percentage = diagSetting->diagSamplingPercentage;
+        return (floor((number - 2) * percentage / 100.0) < floor((number - 1) * percentage / 100.0));
```
Comment (ewertons) [inline iothub_client/src/iothub_client_diagnostic.c]: Only one point of return per function (see other comment with details)

File: iothub_client/src/iothub_client_diagnostic.c [hunk 5]
```diff
@@ -0,0 +1,103 @@
...
+            return false;
+        }
+
+        //TODO: get current time with milliseconds precision
+        char timeBuffer[30];
+        get_current_time_utc(timeBuffer, 30, "%Y-%m-%dT%H:%M:%SZ");
+        if (IoTHubMessage_SetDiagnosticCreationTimeUtc(messageHandle, timeBuffer) != IOTHUB_MESSAGE_OK)
+        {
+            /* Codes_SRS_IOTHUB_DIAGNOSTIC_13_002: [ IoTHubClient_Diagnostic_AddIfNecessary should return false if failing to add diagnostic property. ]*/
+            return false;
```
Comment (ewertons) [inline iothub_client/src/iothub_client_diagnostic.c]: Only one point of return per function (see other comment with details)

File: iothub_client/src/iothub_client_ll.c [hunk 6]
```diff
@@ -1575,6 +1587,23 @@ IOTHUB_CLIENT_RESULT IoTHubClient_LL_SetOption(IOTHUB_CLIENT_LL_HANDLE iotHubCli
                 result = IOTHUB_CLIENT_OK;
             }
         }
+        else if (strcmp(optionName, OPTION_DIAGNOSTIC_SAMPLING_PERCENTAGE) == 0)
+        {
+            uint32_t percentage = *(uint32_t*)value;
+            if (percentage < 0 || percentage > 100)
```
Comment (ewertons) [inline iothub_client/src/iothub_client_ll.c]: You should not check for negative, this is an unsigned int.

File: iothub_client/src/iothub_client_ll.c [hunk 7]
```diff
@@ -1575,6 +1587,23 @@ IOTHUB_CLIENT_RESULT IoTHubClient_LL_SetOption(IOTHUB_CLIENT_LL_HANDLE iotHubCli
...
+            {
+                /*Codes_SRS_IOTHUBCLIENT_LL_10_036: [Calling IoTHubClient_LL_SetOption with value > 100 shall return `IOTHUB_CLIENT_ERRROR`. ]*/
+                LogError("The value of diag_sampling_percentage must be [0, 100]");
```
Comment (ewertons) [inline iothub_client/src/iothub_client_ll.c]: The log is mainly used by the SDK team for debugging, so printing the value is useful. Just add "(%d)" at the end of the string and pass the percentage variable.

File: iothub_client/src/iothub_client_ll.c [hunk 8]
```diff
@@ -1575,6 +1587,23 @@ IOTHUB_CLIENT_RESULT IoTHubClient_LL_SetOption(IOTHUB_CLIENT_LL_HANDLE iotHubCli
...
+                result = IOTHUB_CLIENT_ERROR;
+            }
+            else
+            {
+                /*Codes_SRS_IOTHUBCLIENT_LL_10_037: [Calling IoTHubClient_LL_SetOption with value between [0, 100] shall return `IOTHUB_CLIENT_OK`. ]*/
+                handleData->diagnostic_setting.diagSamplingPercentage = (uint32_t)percentage;
```
Comment (ewertons) [inline iothub_client/src/iothub_client_ll.c]: No need to cast.

File: iothub_client/src/iothub_client_diagnostic.c [hunk 9]
```diff
@@ -0,0 +1,103 @@
...
+    {
+        return '0' + value;
+    }
+    else
+    {
+        return 'a' + value - 10;
+    }
+}
+
+static char* generate_eight_random_characters(char *randomString)
```
Comment (ewertons) [inline iothub_client/src/iothub_client_diagnostic.c]: This function doesn't seem to generate true random strings. I believe there is a good chance of collision here.

You should opt to use UUID (RFC 4122). 

There is a module in c-utility (uniqueid.h) that can generate UUID strings.

File: iothub_client/src/iothub_client_diagnostic.c [hunk 10]
```diff
@@ -0,0 +1,103 @@
...
+{
+    char* randomStringPos = randomString;
+    for (int i = 0; i < 4; ++i)
+    {
+        int rawRandom = rand();
```
Comment (ewertons) [inline iothub_client/src/iothub_client_diagnostic.c]: If using UUID is still not preferred, you should really use srand.

File: iothub_client/src/iothub_client_diagnostic.c [hunk 11]
```diff
@@ -0,0 +1,103 @@
...
+    return randomString;
+}
+
+static bool should_add_diagnostic_info(IOTHUB_DIAGNOSTIC_SETTING_DATA* diagSetting)
+{
+    if (diagSetting->diagSamplingPercentage > 0)
+    {
+        if (diagSetting->currentMessageNumber == UINT32_MAX)
+        {
+            diagSetting->currentMessageNumber %= diagSetting->diagSamplingPercentage;
```
Comment (ewertons) [inline iothub_client/src/iothub_client_diagnostic.c]: If currentMessageNumber is about to overflow (to back to 00000...), the value shall be set to the rest of division of (UINT32_MAX) by currentMessageNumber. 

What is the purpose of this?

File: iothub_client/src/iothub_client_diagnostic.c (hunk 4 above)
Comment (ewertons) [inline iothub_client/src/iothub_client_diagnostic.c]: Do you need to divide by 100.0 on both sides?

File: iothub_client/src/iothub_client_diagnostic.c [hunk 12]
```diff
@@ -0,0 +1,103 @@
...
+        ++diagSetting->currentMessageNumber;
+
+        double number = diagSetting->currentMessageNumber;
+        double percentage = diagSetting->diagSamplingPercentage;
+        return (floor((number - 2) * percentage / 100.0) < floor((number - 1) * percentage / 100.0));
+    }
+    return false;
+}
+
+bool IoTHubClient_Diagnostic_AddIfNecessary(IOTHUB_DIAGNOSTIC_SETTING_DATA* diagSetting, IOTHUB_MESSAGE_HANDLE messageHandle)
```
Comment (ewertons) [inline iothub_client/src/iothub_client_diagnostic.c]: This function should return int, and use __FAILURE__ macro.

File: iothub_client/src/iothub_client_diagnostic.c [hunk 13]
```diff
@@ -0,0 +1,103 @@
...
+
+        //generating random string
+        char randomString[9];
+        if (IoTHubMessage_SetDiagnosticId(messageHandle, generate_eight_random_characters(randomString)) != IOTHUB_MESSAGE_OK)
+        {
+            /* Codes_SRS_IOTHUB_DIAGNOSTIC_13_002: [ IoTHubClient_Diagnostic_AddIfNecessary should return false if failing to add diagnostic property. ]*/
+            return false;
+        }
+
+        //TODO: get current time with milliseconds precision
```
Comment (ewertons) [inline iothub_client/src/iothub_client_diagnostic.c]: Please complete and remove all "TODOs"

File: iothub_client/src/iothub_client_ll.c [hunk 14]
```diff
@@ -461,8 +466,8 @@ static IOTHUB_DEVICE_TWIN* dev_twin_data_create(IOTHUB_CLIENT_LL_HANDLE_DATA* ha
             result->ms_timesOutAfter = 0;
             result->context = userContextCallback;
             result->reported_state_callback = reportedStateCallback;
-			result->client_handle = handleData;
-			result->device_handle = handleData->deviceHandle;
+            result->client_handle = handleData;
+            result->device_handle = handleData->deviceHandle;
```
Comment (ewertons) [inline iothub_client/src/iothub_client_ll.c]: Why these changes are included? please revert.

File: iothub_client/src/iothub_message.c [hunk 15]
```diff
@@ -62,6 +64,36 @@ static int ValidateAsciiCharactersFilter(const char* mapKey, const char* mapValu
     return result;
 }
 
+static void InitializeIoTHubMessage(IOTHUB_MESSAGE_HANDLE handleData)
```
Comment (ewertons) [inline iothub_client/src/iothub_message.c]: please remove this function.

File: iothub_client/src/iothub_message.c [hunk 16]
```diff
@@ -213,57 +265,36 @@ IOTHUB_MESSAGE_HANDLE IoTHubMessage_Clone(IOTHUB_MESSAGE_HANDLE iotHubMessageHan
         }
         else
         {
-            result->messageId = NULL;
-            result->correlationId = NULL;
-            result->userDefinedContentType = NULL;
-            result->contentEncoding = NULL;
+            InitializeIoTHubMessage(result);
```
Comment (ewertons) [inline iothub_client/src/iothub_message.c]: If you want to initialize the new fields to 0, just add the two to the list. Better make it explicit in the function

File: iothub_client/src/iothub_message.c [hunk 17]
```diff
@@ -213,57 +265,36 @@ IOTHUB_MESSAGE_HANDLE IoTHubMessage_Clone(IOTHUB_MESSAGE_HANDLE iotHubMessageHan
...
-                    free(result->userDefinedContentType);
-                }
-                free(result);
+                IoTHubMessage_Destroy(result);
+                result = NULL;
+            }
+            else if (source->diagnosticId != NULL && mallocAndStrcpy_s(&result->diagnosticId, source->diagnosticId) != 0)
+            {
+                LogError("unable to copy diagnosticId");
+                IoTHubMessage_Destroy(result);
```
Comment (ewertons) [inline iothub_client/src/iothub_message.c]: Mixed feelings about this...
Thanks for the optimization, it makes it cleaner.
However, this change is not part of the diagnostics, and I rather not see them in this commit.
And usually we don't call public APIs within the code (at least we should not).

File: iothub_client/tests/iothubclient_diagnostic_ut/iothubclient_diagnostic_ut.c [hunk 18]
```diff
@@ -0,0 +1,245 @@
+// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
```
Comment (ewertons) [inline iothub_client/tests/iothubclient_diagnostic_ut/iothubclient_diagnostic_ut.c]: Congrats on getting the usage of umock. Good one.

File: iothub_client/inc/iothub_client_diagnostic.h (hunk 1 above)
Comment (erich-wang) [reply iothub_client/inc/iothub_client_diagnostic.h]: good suggestion

File: iothub_client/src/iothubtransport_mqtt_common.c [hunk 19]
```diff
@@ -76,6 +76,10 @@ static const char* MESSAGE_ID_PROPERTY = "mid";
 static const char* CORRELATION_ID_PROPERTY = "cid";
 static const char* CONTENT_TYPE_PROPERTY = "ct";
 static const char* CONTENT_ENCODING_PROPERTY = "ce";
+static const char* DIAGNOSTIC_ID_PROPERTY = "diagid";
+static const char* DIAGNOSTIC_CONTEXT_PROPERTY = "diagcxt";
```
[the remark this answers is not in the document]
Comment (erich-wang) [reply iothub_client/src/iothubtransport_mqtt_common.c]: good catch

File: iothub_client/src/iothubtransport_mqtt_common.c [hunk 20]
```diff
@@ -93,6 +97,7 @@ typedef struct SYSTEM_PROPERTY_INFO_TAG
     size_t propLength;
 } SYSTEM_PROPERTY_INFO;
 
+//TODO: add diagnostic properties support for C2D messages
```
[the remark this answers is not in the document]
Comment (erich-wang) [reply iothub_client/src/iothubtransport_mqtt_common.c]: diagnostic properties support for C2D messages is P2 task, we have backlog to track this task. How about just leave TODO here so people would know it is on purpose instead of a bug?

File: iothub_client/src/iothub_client_diagnostic.c (hunk 9 above)
Comment (erich-wang) [reply iothub_client/src/iothub_client_diagnostic.c]: UUID was our first proposal, but changed to 8 digit characters after review

File: iothub_client/src/iothub_client_diagnostic.c (hunk 10 above)
Comment (erich-wang) [reply iothub_client/src/iothub_client_diagnostic.c]: add srand in initialize_iothub_client in iothub_client_ll.c

File: iothub_client/src/iothub_client_diagnostic.c (hunk 11 above)
Comment (erich-wang) [reply iothub_client/src/iothub_client_diagnostic.c]: The purpose is for accurate sampling, updated.

File: iothub_client/inc/iothub_client_diagnostic.h [hunk 21]
```diff
@@ -0,0 +1,47 @@
...
+
+/** @brief diagnostic related setting */
+typedef struct IOTHUB_DIAGNOSTIC_SETTING_DATA_TAG
+{
+    uint32_t diagSamplingPercentage;
+    uint32_t currentMessageNumber;
+} IOTHUB_DIAGNOSTIC_SETTING_DATA;
+
+/**
+    * @brief	Add diagnostic information to message if meeting diagnostic rule
```
[the remark this answers is not in the document]
Comment (erich-wang) [reply iothub_client/inc/iothub_client_diagnostic.h]: updated

File: iothub_client/inc/iothub_client_diagnostic.h [hunk 22]
```diff
@@ -0,0 +1,47 @@
...
+    *
+    * @param	diagSetting		Pointer to an @c IOTHUB_DIAGNOSTIC_SETTING_DATA structure
+    *
+    * @param	messageHandle	message handle 
+    *
+    * @return	true upon success or false if any error
```
[the remark this answers is not in the document]
Comment (erich-wang) [reply iothub_client/inc/iothub_client_diagnostic.h]: updated

File: iothub_client/src/iothub_client_diagnostic.c (hunk 4 above)
Comment (erich-wang) [reply iothub_client/src/iothub_client_diagnostic.c]: updated

Comment (erich-wang) [reply iothub_client/src/iothub_client_diagnostic.c]: yes, it is used to make sure sampling messages with diagnostic properties are uniform among all messages, e.g. if diagSamplingPercentage is equal to 20, the messages flow should be :
D N N N N D N N N N D...
D means message with diagnostic properties while N means normal message

File: iothub_client/src/iothub_client_diagnostic.c (hunk 12 above)
Comment (erich-wang) [reply iothub_client/src/iothub_client_diagnostic.c]: updated

File: iothub_client/src/iothub_client_diagnostic.c (hunk 13 above)
Comment (erich-wang) [reply iothub_client/src/iothub_client_diagnostic.c]: done

File: iothub_client/src/iothub_client_diagnostic.c (hunk 5 above)
Comment (erich-wang) [reply iothub_client/src/iothub_client_diagnostic.c]: done

File: iothub_client/inc/iothub_client_options.h (hunk 2 above)
Comment (erich-wang) [reply iothub_client/inc/iothub_client_options.h]: yes, finally will be applicable to all kinds of messages.

File: iothub_client/src/iothub_client_diagnostic.c (hunk 3 above)
Comment (erich-wang) [reply iothub_client/src/iothub_client_diagnostic.c]: thanks for the detail guidelines

File: iothub_client/src/iothub_client_ll.c (hunk 14 above)
Comment (erich-wang) [reply iothub_client/src/iothub_client_ll.c]: The change must be introduced by formatting the file.

It was updated with **correct** indent spaces, do you want me to revert it and use separate PR to commit the change?

File: iothub_client/src/iothub_client_ll.c (hunk 6 above)
Comment (erich-wang) [reply iothub_client/src/iothub_client_ll.c]: done

File: iothub_client/src/iothub_client_ll.c (hunk 8 above)
Comment (erich-wang) [reply iothub_client/src/iothub_client_ll.c]: done

File: iothub_client/src/iothub_message.c (hunk 17 above)
Comment (erich-wang) [reply iothub_client/src/iothub_message.c]: ok, let's refactoring the code in separate commit.

File: iothub_client/src/iothub_message.c (hunk 16 above)
Comment (erich-wang) [reply iothub_client/src/iothub_message.c]: done

File: iothub_client/src/iothub_message.c (hunk 15 above)
Comment (erich-wang) [reply iothub_client/src/iothub_message.c]: done

File: iothub_client/src/iothub_client_ll.c (hunk 7 above)
Comment (erich-wang) [reply iothub_client/src/iothub_client_ll.c]: updated

File: iothub_client/src/iothubtransport_mqtt_common.c (hunk 20 above)
[the remark this answers is not in the document]
Comment (ewertons) [reply iothub_client/src/iothubtransport_mqtt_common.c]: I believe we will have trouble if this TODO tags get into the repo.
@jebrando once told me that our code is inspected by some group at Microsoft that explicitly hunts for TODO statements... 
he might be able to give more details.

File: iothub_client/src/iothubtransport_mqtt_common.c [hunk 23]
```diff
@@ -639,6 +644,75 @@ static STRING_HANDLE addPropertiesTouMqttMessage(IOTHUB_MESSAGE_HANDLE iothub_me
...
+                    STRING_HANDLE diagContextHandle = STRING_construct_sprintf("%s=%s", DIAGNOSTIC_CONTEXT_CREATION_TIME_UTC_PROPERTY, creation_time_utc);
+                    if (diagContextHandle == NULL)
+                    {
+                        LogError("Failed constructing diagnostic context");
+                        STRING_delete(result);
+                        result = NULL;
+                    }
+                    //Add other diagnostic context properties here if have more
+
+                    if (diagContextHandle != NULL)
```
Comment (ewertons) [inline iothub_client/src/iothubtransport_mqtt_common.c]: This seems incorrect. It should be "else", you don't need to test the opposite value again.

Also, is the comment above intentional?

File: iothub_client/src/iothubtransport_mqtt_common.c [hunk 24]
```diff
@@ -639,6 +644,75 @@ static STRING_HANDLE addPropertiesTouMqttMessage(IOTHUB_MESSAGE_HANDLE iothub_me
...
+                        }
+                        STRING_delete(diagContextHandle);
+                        diagContextHandle = NULL;
+                        index++;
+                    }
+                }
+            }
+            else if (diag_id != NULL || creation_time_utc != NULL)
+            {
+                // Codes_SRS_IOTHUB_TRANSPORT_MQTT_COMMON_09_015: [ `IoTHubTransport_MQTT_Common_DoWork` shall check whether diagid and diagCreationTimeUtc be present simultaneously, trade as error if not]
```
Comment (ewertons) [inline iothub_client/src/iothubtransport_mqtt_common.c]: Typo: "trade".
Should be "treat"?

File: iothub_client/src/iothubtransport_mqtt_common.c (hunk 20 above)
[the remark this answers is not in the document]
Comment (erich-wang) [reply iothub_client/src/iothubtransport_mqtt_common.c]: TODO removed.

File: iothub_client/src/iothubtransport_mqtt_common.c (hunk 23 above)
Comment (erich-wang) [reply iothub_client/src/iothubtransport_mqtt_common.c]: yes, the comment is intentional. Currently there's only one property within diagnostic context, the comments is telling where to add code if having more diagnostic properties.

File: iothub_client/src/iothubtransport_mqtt_common.c (hunk 24 above)
Comment (erich-wang) [reply iothub_client/src/iothubtransport_mqtt_common.c]: done
